#### Name Dipika Sharma
#### DSC 550 T302 Data Mining
#### Week 8

## Exercise 8.2

### Best Model Selection and Hyperparameter Tuning

In [1]:
## Lets import the reqired libraries.

import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split #used to split data into training/test sets

## from sklearn.metrics import plot_confusion_matrix
from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay
import warnings
warnings.filterwarnings("ignore")
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

### 1. Import the dataset and ensure that it loaded properly.

In [2]:
## Lets import the dataset Loan_Train.

Loan_Train_df = pd.read_csv('Loan_Train.csv') 

In [3]:
## Lets check first five rows of the data frame.

Loan_Train_df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

### 2. Prepare the data for modeling by performing the following steps: 
#### &emsp;&emsp; Drop the column “Loan_ID.” <br> &emsp;&emsp; Drop any rows with missing data.<br> &emsp;&emsp; Convert the categorical features into dummy variables.

In [4]:
## Lets check how many columns and rows we have in data frame.

Loan_Train_df.shape

(614, 13)

In [5]:
## Lets drop the column Load_ID from the data frame.

Loan_Train_df = Loan_Train_df.drop(['Loan_ID'], axis=1)

In [6]:
## Lets again check how many columns and rows we have in data frame.

Loan_Train_df.shape

(614, 12)

#### We can see the number of rows reduce from 13 to 12 after dropping Loan_ID colulmn from the data frame.

In [7]:
## Using dropna function to drop any rows with missing data in data frame.

Loan_Train_df = Loan_Train_df.dropna()

In [8]:
## Lets again check how many columns and rows we have in data frame.

Loan_Train_df.shape

(480, 12)

#### As we can see after dropping rows with missing values from the data fram the number rows went down from 614 to 480 rows.

In [9]:
## Now lets convert the categorical features of the data frame into the dummy variables.
## Lets find out the categorical columns in the dataframe.

CateColumns = Loan_Train_df.select_dtypes("object").columns

## Storing the categorical columns in a list.

CateColumns= list(set(CateColumns))

In [10]:
## Lets convert the categorical column Loan_status to numerical.
## We know if Class status is "Y" then code is 1 and if it is "N" then the code is 0.

Loan_Train_df['Loan_Status_Code'] = Loan_Train_df['Loan_Status'].replace(to_replace=["Y","N"], value=[1,0])

In [11]:
## Lets print out the categorical columns of the dataframe.

CateColumns

['Gender',
 'Education',
 'Property_Area',
 'Dependents',
 'Married',
 'Loan_Status',
 'Self_Employed']

In [12]:
## Using the get_dummies function for creating the dummmy variables.

Loan_Train_df = pd.get_dummies(Loan_Train_df, columns=CateColumns)

In [13]:
## Lets display the top 5 rows of the dataframe.

Loan_Train_df.head()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   
5             5417             4196.0       267.0             360.0   

   Credit_History  Loan_Status_Code  Gender_Female  Gender_Male  \
1             1.0                 0              0            1   
2             1.0                 1              0            1   
3             1.0                 1              0            1   
4             1.0                 1              0            1   
5             1.0                 1              0            1   

   Education_Graduate  Education_Not Graduate  ...  Dependents_0  \
1                   1                       0  ...             0   
2                   1                       0  ...             1   
3                   0                       1  ...             1   
4                   1                       0  ...             1   
5                   1                       0  ...             0   

   Dependents_1  Dependents_2  Dependents_3+  Married_No  Married_Yes  \
1             1             0              0           0            1   
2             0             0              0           0            1   
3             0             0              0           0            1   
4             0             0              0           1            0   
5             0             1              0           0            1   

   Loan_Status_N  Loan_Status_Y  Self_Employed_No  Self_Employed_Yes  
1              1              0                 1                  0  
2              0              1                 0                  1  
3              0              1                 1                  0  
4              0              1                 1                  0  
5              0              1                 0                  1  

[5 rows x 23 columns]

In [14]:
## Lets check out the number of rows and columns of the new dataframe.

Loan_Train_df.shape

(480, 23)

#### As we can see after adding the dummy variables the number of column increases from 12 to 23.

### 3. Split the data into a training and test set, where the “Loan_Status” column is the target.

In [15]:
## Lets split the data into a training and test set.

X = Loan_Train_df.drop(['Loan_Status_Code','Loan_Status_N','Loan_Status_Y'], axis = 1)

Y = Loan_Train_df['Loan_Status_Code']

In [16]:
## Pareto Principle Split
## Using train_test_split function to split the data into 80% training and 20% testing.

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size= 0.2)

In [17]:
## Lets reset the index in the training and test sets.

X_train = X_train.reset_index(drop = True)
X_test = X_test.reset_index(drop = True)
Y_train = Y_train.reset_index(drop = True)
Y_test = Y_test.reset_index(drop = True)

In [18]:
## lets check the number of rows and columns of the split data

print("The shape of X training dataset is: {}".format(X_train.shape))
print("The shape of Y training dataset is: {}".format(Y_train.shape))
print("The shape of X testing dataset is: {}".format(X_test.shape))
print("The shape of Y testing dataset is: {}".format(Y_test.shape))

The shape of X training dataset is: (384, 20)
The shape of Y training dataset is: (384,)
The shape of X testing dataset is: (96, 20)
The shape of Y testing dataset is: (96,)


In [19]:
## Lets see how many values of saleprice are in training and test sets

print(Y_train.value_counts())
print(Y_test.value_counts())

1    267
0    117
Name: Loan_Status_Code, dtype: int64
1    65
0    31
Name: Loan_Status_Code, dtype: int64


In [20]:
## Lets look for nulls in both training and test sets.

print(X_train.isna().sum())
print(X_test.isna().sum())

ApplicantIncome            0
CoapplicantIncome          0
LoanAmount                 0
Loan_Amount_Term           0
Credit_History             0
Gender_Female              0
Gender_Male                0
Education_Graduate         0
Education_Not Graduate     0
Property_Area_Rural        0
Property_Area_Semiurban    0
Property_Area_Urban        0
Dependents_0               0
Dependents_1               0
Dependents_2               0
Dependents_3+              0
Married_No                 0
Married_Yes                0
Self_Employed_No           0
Self_Employed_Yes          0
dtype: int64
ApplicantIncome            0
CoapplicantIncome          0
LoanAmount                 0
Loan_Amount_Term           0
Credit_History             0
Gender_Female              0
Gender_Male                0
Education_Graduate         0
Education_Not Graduate     0
Property_Area_Rural        0
Property_Area_Semiurban    0
Property_Area_Urban        0
Dependents_0               0
Dependents_1               0
D

### 4. Create a pipeline with a min-max scaler and a KNN classifier (see section 15.3 in the Machine Learning with Python Cookbook).

In [21]:
## Lets import the required libraries

from sklearn.neighbors import KNeighborsClassifier
from sklearn import datasets
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.model_selection import GridSearchCV
from sklearn import preprocessing

In [22]:
## Lets initialize the minmax scaler

minmax_scale = preprocessing.MinMaxScaler(feature_range=(0,1))

In [23]:
## Lets initialize the KNN Classifier

knn = KNeighborsClassifier(n_neighbors=5, n_jobs=-1)

In [24]:
## Creating the pipeline with min-max scaler and a KNN classifier

pipe_line = Pipeline([("scaler", minmax_scale), ("knn", knn)])

### 5. Fit a default KNN classifier to the data with this pipeline. Report the model accuracy on the test set. Note: Fitting a pipeline model works just like fitting a regular model.

In [25]:
## Lets fit a default KNN classifier to the data with the pipeline

pipe_line.fit(X_train, Y_train) 

Pipeline(steps=[('scaler', MinMaxScaler()),
                ('knn', KNeighborsClassifier(n_jobs=-1))])

In [26]:
## Making prediction on test set

Y_pred = pipe_line.predict(X_test)
Y_pred

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1])

In [27]:
## Calculating the model accuracy on the test set.

pipe_line.score(X_test,Y_test)

0.6875

### 6. Create a search space for your KNN classifier where your “n_neighbors” parameter varies from 1 to 10. (see section 15.3 in the Machine Learning with Python Cookbook).

In [28]:
## Create a search space for KNN classifier.

SSpace = [{"knn__n_neighbors": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}]

### 7. Fit a grid search with your pipeline, search space, and 5-fold cross-validation to find the best value for the “n_neighbors” parameter.

In [29]:
## Lets instantiate the StandardScaler visualizer

StndScaler = StandardScaler()

In [30]:
## Lets fit the visualizer and the model

Matrix_X = StndScaler.fit_transform(X)

In [31]:
## Lets do grod search with pipeline.

classifier = GridSearchCV(pipe_line, SSpace, cv=5, verbose=0).fit(Matrix_X, Y)

In [32]:
## Lets find best value for the “n_neighbors” parameter.

classifier.best_estimator_.get_params()["knn__n_neighbors"]

3

### 8. Find the accuracy of the grid search best model on the test set. Note: It is possible that this will not be an improvement over the default model, but likely it will be.

In [33]:
print("The best parameters for the KNN tuned hpyerparameters is : {}".format(classifier.best_params_))
print("The accuracy of the grid search best model on the test set is : ", round(classifier.best_score_*100, 2), '%', sep = '')

The best parameters for the KNN tuned hpyerparameters is : {'knn__n_neighbors': 3}
The accuracy of the grid search best model on the test set is : 72.29%


### 9. Now, repeat steps 6 and 7 with the same pipeline, but expand your search space to include logistic regression and random forest models with the hyperparameter values in section 12.3 of the Machine Learning with Python Cookbook.

In [34]:
## Repeating steps 6 and 7 - pipeline with KNN hyperparameters

pipe_line = Pipeline([('StndScaler', StandardScaler()), ("classifier", KNeighborsClassifier(n_neighbors=5, n_jobs=-1))])

In [35]:
## Expand search space to include logistic regression and random forest models

SSpace=[
    {'classifier' : [KNeighborsClassifier()], 
     'classifier__n_neighbors' : [1,2,3,4,5,6,7,8,9,10]},
    {'classifier' : [LogisticRegression()],
     'classifier__penalty' : ['l1', 'l2'],
    'classifier__C' : np.logspace(0, 4, 10),
    'classifier__solver' : ['liblinear']},
    {'classifier' : [RandomForestClassifier()],
    'classifier__n_estimators' : [10,100,1000],
    'classifier__max_features' : [1, 2, 3]}
]


In [36]:
## Creating grid search

GSearch = GridSearchCV(pipe_line, SSpace, cv = 5, verbose=0)

In [37]:
## Lets fir the grid search.

BModel = GSearch.fit(X_train, Y_train)

### 10. What are the best model and hyperparameters found in the grid search? Find the accuracy of this model on the test set.

In [38]:
## Lets find out the best model and hyperparameters in grid search.

BModel.best_estimator_.get_params()

{'memory': None,
 'steps': [('StndScaler', StandardScaler()),
  ('classifier', LogisticRegression(penalty='l1', solver='liblinear'))],
 'verbose': False,
 'StndScaler': StandardScaler(),
 'classifier': LogisticRegression(penalty='l1', solver='liblinear'),
 'StndScaler__copy': True,
 'StndScaler__with_mean': True,
 'StndScaler__with_std': True,
 'classifier__C': 1.0,
 'classifier__class_weight': None,
 'classifier__dual': False,
 'classifier__fit_intercept': True,
 'classifier__intercept_scaling': 1,
 'classifier__l1_ratio': None,
 'classifier__max_iter': 100,
 'classifier__multi_class': 'auto',
 'classifier__n_jobs': None,
 'classifier__penalty': 'l1',
 'classifier__random_state': None,
 'classifier__solver': 'liblinear',
 'classifier__tol': 0.0001,
 'classifier__verbose': 0,
 'classifier__warm_start': False}

In [39]:
## Lets fit grid search.

BModel=GSearch.fit(X_test, Y_test)

In [42]:
print("The accuracy of Best Model is:",BModel.best_score_)

The accuracy of Best Model is: 0.76


### 11. Summarize your results.

#### The Best Model to use with hyperparameters on Loan_Train dataset is Logistic Regression with 76% accuracy.